### PR TITLE
Routes List

### PR DESCRIPTION
**Pull request by SMenigat**
PR ist schonmal offen, da sich an dem Routes Endpoint in der REST-API noch einiges ändern muss. Unsere Anforderungen mit Sprachen, zusätzlich zugelassenen Pionieren etc. sind noch nicht implementiert.

Das muss dann nachträglich alles hinzugefügt werden denke ich. Zudem sind hier schon Referenzen auf Sprache und Depot enthalten. Wäre also besser wenn man diese erstmal voll pflegbar macht bevor man hier weitermacht.

### CODE REVIEW

File: app/containers/App/index.js [hunk 1]
```diff
@@ -15,8 +15,13 @@ import React, { PropTypes } from 'react';
...
 import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
-import Drawer from 'material-ui/Drawer';
-import MenuItem from 'material-ui/MenuItem';
+import {
+  Divider,
+  Drawer,
+  FontIcon,
+  MenuItem,
+} from 'material-ui';
+import { browserHistory } from 'react-router';
```
Comment (PhilippSpo) [inline app/containers/App/index.js]: Wir sollten die `browserHistory` wenn es geht nicht direkt nutzen sondern die `push` methode auf der `router` instanz aufrufen.
Wenn du die Component in der [`withRouter()`] HOC wrapst bekommst du das `router` Objekt als ein prop rein.
```js
import { withRouter } from 'react-router'
// ...
gotoRoute = (route) => {
  this.props.router.push(route);
}
// ...
export default withRouter(connect(
  mapStateToProps,
  {
    logout: actions.logout,
    setDrawerMenuState: actions.setDrawerMenuState,
  }
)(App));
```

File: app/containers/App/index.js [hunk 2]
```diff
@@ -34,6 +39,11 @@ export class App extends React.Component {
     setDrawerMenuState: PropTypes.func.isRequired,
   };
 
+  gotoRoute = (route) => {
+    this.toggleDrawerMenuState();
```
Comment (PhilippSpo) [inline app/containers/App/index.js]: Hat jetzt mit dem PR direkt nix zu tun, aber die methode sollten wir eher `toggleDrawerMenu` nennen.

File: app/containers/App/sagas.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ export function* login({ user, password, redirectTo = '/' }) {
     const token = yield call(
       callApi,
       {
-        endpoint: 'authenticate',
+        endpoint: 'authenticate/forms',
```
Comment (PhilippSpo) [inline app/containers/App/sagas.js]: Was ist das für ein endpoint? Warum brauchen wir das `/forms`? Vtl ein Kommentar in den Code hier.

File: app/containers/RouteManagement/constants.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export const ROUTE_REQUEST = 'ROUTE_REQUEST';
+export const ROUTE_REQUESTED = 'ROUTE_REQUESTED';
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/constants.js]: den brauchen wir nimmer

File: app/containers/RouteManagement/index.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          title="Route Management"
+          onLeftIconButtonTouchTap={
+            () => this.props.setDrawerMenuState({ isDrawerMenuOpen: true })
+          }
+        />
+        {this.props.isFetching ? (
+          <CircularProgress size={60} thickness={7} />
+        ) : (
+          <List>
+            {routes.map((route) => <ListItem
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/index.js]: Bei multiline react elements bitte einklammern:
```js
{routes.map((route) => (
  <ListItem
    primaryText={route.name}
    secondaryText={`On site location: ${route.onSiteLocation}`}
    key={route.id}
  />
))}
```

File: app/containers/RouteManagement/tests/index.tests.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+describe('<RouteManagement />', () => {
+  describe('rendering', () => {
+    it('should render AppBar', () => {
+      const props = getProps();
+      const wrapper = shallow(
+        <RouteManagement {...props} />
+      );
+      expect(wrapper.find('AppBar').length).toEqual(1);
+    });
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/tests/index.tests.js]: Hier fehlen noch tests für das rendern des loaderns bzw. der routes, sowie dem aufrufen von `loadRoutes`.

File: app/reducers.js [hunk 7]
```diff
@@ -45,6 +46,7 @@ function routeReducer(state = routeInitialState, action) {
 export default function createReducer(asyncReducers) {
   return combineReducers({
     route: routeReducer,
+    routes: routesReducer,
```
Comment (PhilippSpo) [inline app/reducers.js]: wir sollten die "dynamischen" reducer (also solche die nicht auf jeder page der app gebraucht werden) nicht hier einbauen, weil sie dann immer geladen werden und wir uns somit code-splitting kaputt machen.
[Wir sollten die dynamisch nachladen wenn man auf die route geht in der route definition.](https://github.com/mxstbr/react-boilerplate/blob/master/docs/js/routing.md#dynamic-routes)

**Comment by SMenigat** [comment]
Hab den PR direkt mal aktualisiert. Die Änderungsvorschläge sind umgesetzt. Die fehlenden Tests für Reducer, Actions und Selectors wurden hinzugefügt. Zudem habe ich den slice nochmal refactored. Die route Objekte werden jetzt nicht mehr in einem Array sondern in einer Map nach Id gespeichert.

File: app/containers/App/tests/sagas.test.js [hunk 8]
```diff
@@ -148,7 +148,7 @@ describe('apiWatcher Saga', () => {
   });
   it('should call the `callApi`', () => {
     expect(gen.next().value).toEqual(
-      call(callApi, { endpoint: 'routes', authenticated: true })
+      call(callApi, { endpoint: 'routes', authenticated: true, config: undefined })
```
Comment (PhilippSpo) [inline app/containers/App/tests/sagas.test.js]: isn't this the same as not passing `config` at all?

File: app/containers/RouteManagement/index.js [hunk 9]
```diff
@@ -56,7 +61,7 @@ export class RouteManagement extends React.Component { // eslint-disable-line re
 
 const mapStateToProps = createStructuredSelector({
   isFetching: selectIsFetching(),
-  routes: selectRoutes(),
+  routesById: selectRoutesById(),
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/index.js]: selectRoutes sollte einfach nur ein Array an routes zurück geben, keine hash map. Dann muss es auch nicht `routesById` heißen sondern kann immernoch `routes` heißen. Routes management sollte nichts darüber wissen wie der redux state aussieht.

File: app/containers/RouteManagement/tests/index.test.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      );
+      expect(props.loadRoutes.calledOnce).toEqual(true);
+    });
+    it('should render AppBar', () => {
+      const props = getProps();
+      const wrapper = shallow(
+        <RouteManagement {...props} />
+      );
+      expect(wrapper.find('AppBar').length).toEqual(1);
+    });
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/tests/index.test.js]: hier sollten wir auch testen dass die App bar auch den drawer state toggled.

File: app/containers/App/tests/sagas.test.js (hunk 8 above)
Comment (SMenigat) [inline app/containers/App/tests/sagas.test.js]: Nope. Not passing config as undefined, causes the test to fail:

Expected { '@@redux-saga/IO': true, CALL: { args: [ { authenticated: true, config: undefined, endpoint: 'routes' } ], context: null, fn: [Function: callApi] } } to equal { '@@redux-saga/IO': true, CALL: { args: [ { authenticated: true, endpoint: 'routes' } ], context: null, fn: [Function: callApi] } }

Comment (PhilippSpo) [inline app/containers/App/tests/sagas.test.js]: können wir nicht einfach den test anpassen?

File: app/containers/RouteManagement/tests/index.test.js [hunk 11]
```diff
@@ -23,22 +25,37 @@ const mountWithContext = (node) => mount(node, {
...
+        expect(wrapper.find('AppBar').length).toEqual(1);
+      });
+      it('should toggle drawer onLeftIconButtonTouchTap', () => {
+        const props = getProps();
+        const wrapper = mountWithContext(
+          <RouteManagement {...props} />
+        );
+        const burgerButton = wrapper.find('AppBar').find('IconButton').first();
+        // to simulate touch tap, we need to hack a bit
+        const node = ReactDOM.findDOMNode(burgerButton.node); // eslint-disable-line
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/tests/index.test.js]: Das kannst alles mit enzyme machen.

File: app/containers/RouteManagement/tests/index.test.js [hunk 12]
```diff
@@ -16,23 +12,11 @@ const getProps = () => ({
   routes: [],
 });
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/tests/index.test.js]: mit context müssen wir nur rendern, wenn wir `mount` nutzen, weil dann die Material UI components auch gerendert werden. Bei shallow rendern wir nur eine Ebene.

File: app/containers/RouteManagement/tests/index.test.js [hunk 13]
```diff
@@ -47,13 +31,10 @@ describe('<RouteManagement />', () => {
...
         const props = getProps();
-        const wrapper = mountWithContext(
+        const wrapper = shallow(
           <RouteManagement {...props} />
         );
-        const burgerButton = wrapper.find('AppBar').find('IconButton').first();
-        // to simulate touch tap, we need to hack a bit
-        const node = ReactDOM.findDOMNode(burgerButton.node); // eslint-disable-line
-        ReactTestUtils.Simulate.touchTap(node);
+        wrapper.find('AppBar').prop('onLeftIconButtonTouchTap')();
```
Comment (PhilippSpo) [inline app/containers/RouteManagement/tests/index.test.js]: @SMenigat Wir müssen hier nicht testen, dass die `AppBar` bei einem Klick auf den `IconButton` unseren Callback aufruft. Das macht der test für die `AppBar` von Material UI. Wir müssen nur testen, dass wenn unser callback in `onLeftIconButtonTouchTap ` aufregufen wird auch die `setDrawerMenuState` action aufgerufen wird.

File: internals/tests/setup.js [hunk 14]
```diff
@@ -1,20 +1,23 @@
...
-    },
-    clear() {
-      store = {};
-    },
-  };
-}());
+import injectTapEventPlugin from 'react-tap-event-plugin';
+
+// enable tap react events
+injectTapEventPlugin();
```
Comment (PhilippSpo) [inline internals/tests/setup.js]: Hier ist das setup file für die Tests. Wir könnten das im moment sagar wieder rausnehmen eigentlich. Kann aber sein dass wir in einem edge-case doch mal `mount` nutzen und dann brauchen wir das inject hier.

**Comment by PhilippSpo** [comment]
@SMenigat was ist jetzt der Scope vom dem PR? Nur eine Liste von Routen anzeigen?

File: app/containers/RouteManagement/tests/index.test.js (hunk 12 above)
Comment (SMenigat) [inline app/containers/RouteManagement/tests/index.test.js]: @PhilippSpo Hatte `mount` nur genutzt weil ich dachte, dass sonst die Mount-Hooks nicht triggern 😓

**Comment by SMenigat** [comment]
Bisher, Ja. Dachte es legt dann schon einmal eine Grundlage auf dem man aufbauen kann, wenn man Add, Edit, Delete implementiert.

**Comment by PhilippSpo** [comment]
OK dann lass den PR umbenennen und mergen ✨
